### PR TITLE
ParallelUtility rethrow OperationCanceledException

### LINKED ISSUE #3510: Parallel Foreach eats TaskCancelled Exception
https://ci.appveyor.com/project/DocFXService/docfx/builds/19606644

### PR DESCRIPTION
**Pull request by yufeih**
Fixes #3510

### CODE REVIEW

File: src/docfx/lib/ParallelUtility.cs [hunk 1]
```diff
@@ -51,19 +60,38 @@ public static Task ForEach<T>(IEnumerable<T> source, Func<T, Task> action, Actio
...
+            }
 
             async Task Run(T item)
             {
-                await action(item);
+                try
+                {
+                    await action(item);
+                }
+                catch (TaskCanceledException tce)
```
Comment (OsmondJiang) [inline src/docfx/lib/ParallelUtility.cs]: do we have any document can reference?

File: src/docfx/lib/ParallelUtility.cs [hunk 2]
```diff
@@ -134,5 +173,11 @@ private static void DebugAssertPostedOrFaulted<T>(bool posted, ActionBlock<T> qu
                 Debug.Assert(queue.Completion.IsFaulted);
             }
         }
+
+        private class WrapException : Exception
```
Comment (OsmondJiang) [inline src/docfx/lib/ParallelUtility.cs]: >WrapException [](start = 22, length = 13)

add link or comment to here, explain why do you do that?

File: src/docfx/lib/ParallelUtility.cs [hunk 3]
```diff
@@ -51,19 +60,34 @@ public static Task ForEach<T>(IEnumerable<T> source, Func<T, Task> action, Actio
...
+            }
 
             async Task Run(T item)
             {
-                await action(item);
+                try
+                {
+                    await action(item);
+                }
+                catch (OperationCanceledException oce)
```
Comment (live1206) [inline src/docfx/lib/ParallelUtility.cs]: >OperationCanceledException [](start = 23, length = 26)

Shall we try a few retry here when we get OperationCanceledException? Especially while we get it with restoring timeout, maybe we do not need to fail the entire run

Comment (yufeih) [reply src/docfx/lib/ParallelUtility.cs]: Retry should be handled in the inner action.

Comment (live1206) [reply src/docfx/lib/ParallelUtility.cs]: true

---
In reply to: [226557534](https://github.com/dotnet/docfx/pull/3514#discussion_r226557534) [](ancestors = 226557534)

**Pull request merged**